oci/layout/oci_dest.go

```diff
@@ -302,6 +303,90 @@ func (d *ociImageDestination) CommitWithOptions(ctx context.Context, options pri
 	return os.WriteFile(d.ref.indexPath(), indexJSON, 0644)
 }
 
+// tryHardlinkLocalFile attempts to hardlink the specified file and digest it.
+func tryHardlinkLocalFile(dest *ociImageDestination, file string) (private.UploadedBlob, bool, error) {
+	blobFile, err := os.CreateTemp(dest.ref.dir, "oci-put-blob")
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+
+	// Remove the file for os.Link() to work.
+	blobFile.Close()
+	os.Remove(blobFile.Name())
+	if err := os.Link(file, blobFile.Name()); err != nil {
+		// Fallback iff Link has failed.
+		return private.UploadedBlob{}, true, err
+	}
+
+	blobFile, err = os.Open(blobFile.Name())
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+	blobDigest, err := digest.FromReader(blobFile)
+	if err != nil {
+		blobFile.Close()
+		return private.UploadedBlob{}, false, err
+	}
+	blobPath, err := dest.ref.blobPath(blobDigest, dest.sharedBlobDir)
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+	if err := ensureParentDirectoryExists(blobPath); err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+
+	// need to explicitly close the file, since a rename won't otherwise work on Windows
+	blobFile.Close()
+	if err := os.Rename(blobFile.Name(), blobPath); err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+
+	fileInfo, err := os.Stat(blobPath)
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+	return private.UploadedBlob{Digest: blobDigest, Size: fileInfo.Size()}, false, nil
+}
+
+// PutBlobFromLocalFile arranges the data from path to be used as blob with digest.
+// It computes, and returns, the digest and size of the used file.
+//
+// This function can be used instead of dest.PutBlob() where the ImageDestination requires PutBlob() to be called.
+func PutBlobFromLocalFile(ctx context.Context, dest types.ImageDestination, file string) (digest.Digest, int64, error) {
```

Comment:
@mtrmac @baude
I think we should add the functionality of specifying the media type of the blob/layer here.  A `podman artifact add --mediatype something.something.ai` may actually be needed.  Otherwise, the manifest would suggest that the layers/files are compressed tar balls.

Reply:
Ah, scratch that.  The caller is expected to update the manifest in those cases.

I implemented a very similar mechanism in `buildah source add` which automatically adds the layers to the manifest and updates the digests etc:
https://github.com/containers/buildah/blob/main/internal/source/add.go#L46

I am OK with the current design but it's probably something we should add to the docs of the function.

Reply:
> Ah, scratch that. The caller is expected to update the manifest in those cases.

Yes. Copy&pasting from https://github.com/containers/image/pull/2567#discussion_r1842779729:

The way the `ImageDestination` API works, `PutBlob` doesn’t typically need to worry about the MIME type: the MIME type comes via `PutManifest`.

While it is true that the `PutBlobWithOptions` gets a full `types.BlobInfo`, and https://github.com/containers/image/blob/0c42f2fe704c9eb84a0b91f4f0363c9919a677db/copy/single.go#L852 forwards the MIME type value into the layer copy code, there are things like https://github.com/containers/image/blob/0c42f2fe704c9eb84a0b91f4f0363c9919a677db/copy/compression.go#L168 just dropping that; so, transports can’t rely on a MIME type being present (~instead, we have `PutBlobOptions.IsConfig`).

Also, _specifically in the OCI transport_, the `PutBlob*` code really doesn’t care about the MIME type. It is going to create a sha256-named file, and as far as the layout definition is concerned, it’s perfectly valid to refer to a single blob in two different manifests using two different MIME types. So, for this _OCI-transport-specific_ function, I don’t think adding a MIME type is right.